fdbserver/workloads/AtomicOps.actor.cpp

```diff
@@ -172,29 +182,138 @@ struct AtomicOpsWorkload : TestWorkload {
+			uint64_t intValue = 0;
+			memcpy(&intValue, kv.value.begin(), kv.value.size());
+			sum += intValue;
+			TraceEvent("AtomicOpLog")
+			    .detail("Key", kv.key)
+			    .detail("Val", kv.value)
+			    .detail("IntValue", intValue)
+			    .detail("CurSum", sum);
+		}
+		return Void();
+	}
+
+	ACTOR Future<Void> dumpDebugKV(Database cx, int g) {
+		ReadYourWritesTransaction tr(cx);
+		Key begin(format("debug%08x", g));
+		Standalone<RangeResultRef> log = wait(tr.getRange(KeyRangeRef(begin, strinc(begin)), CLIENT_KNOBS->TOO_MANY));
+		for (auto& kv : log) {
+			TraceEvent("AtomicOpDebug").detail("Key", kv.key).detail("Val", kv.value);
+		}
+		return Void();
+	}
+
+	ACTOR Future<Void> dumpOpsKV(Database cx, int g) {
+		ReadYourWritesTransaction tr(cx);
+		Key begin(format("ops%08x", g));
+		Standalone<RangeResultRef> ops = wait(tr.getRange(KeyRangeRef(begin, strinc(begin)), CLIENT_KNOBS->TOO_MANY));
+		uint64_t sum = 0;
+		for (auto& kv : ops) {
+			uint64_t intValue = 0;
+			memcpy(&intValue, kv.value.begin(), kv.value.size());
+			sum += intValue;
+			TraceEvent("AtomicOpOps")
+			    .detail("Key", kv.key)
+			    .detail("Val", kv.value)
+			    .detail("IntVal", intValue)
+			    .detail("CurSum", sum);
+		}
+		return Void();
+	}
+
+	ACTOR Future<Void> validateOpsKey(Database cx, AtomicOpsWorkload* self, int g) {
+		// Get mapping between opsKeys and debugKeys
+		state ReadYourWritesTransaction tr1(cx);
+		state std::map<Key, Key> records; // <ops, debugKey>
+		Key begin(format("debug%08x", g));
+		Standalone<RangeResultRef> debuglog =
+		    wait(tr1.getRange(KeyRangeRef(begin, strinc(begin)), CLIENT_KNOBS->TOO_MANY));
+		for (auto& kv : debuglog) {
+			records[kv.value] = kv.key;
+		}
+
+		// Get log key's value and assign it to the associated debugKey
+		state ReadYourWritesTransaction tr2(cx);
+		state std::map<Key, int64_t> logVal; // debugKey, log's value
+		Key begin(format("log%08x", g));
+		Standalone<RangeResultRef> log = wait(tr2.getRange(KeyRangeRef(begin, strinc(begin)), CLIENT_KNOBS->TOO_MANY));
+		for (auto& kv : log) {
+			uint64_t intValue = 0;
+			memcpy(&intValue, kv.value.begin(), kv.value.size());
+			logVal[kv.key.removePrefix(LiteralStringRef("log")).withPrefix(LiteralStringRef("debug"))] = intValue;
+		}
+
+		// Get opsKeys and validate if it has correct value
+		state ReadYourWritesTransaction tr3(cx);
+		state std::map<Key, int64_t> opsVal; // ops key, ops value
+		Key begin(format("ops%08x", g));
+		Standalone<RangeResultRef> ops = wait(tr3.getRange(KeyRangeRef(begin, strinc(begin)), CLIENT_KNOBS->TOO_MANY));
+		// Validate if ops' key value is consistent with logs' key value
+		for (auto& kv : ops) {
+			bool inRecord = records.find(kv.key) != records.end();
+			uint64_t intValue = 0;
+			memcpy(&intValue, kv.value.begin(), kv.value.size());
+			opsVal[kv.key] = intValue;
+			if (!inRecord) {
+				TraceEvent(SevError, "MissingLogKey").detail("OpsKey", kv.key);
+			}
+			if (inRecord && intValue == 0) {
```

Comment:
A key's value is set up as `StringRef((const uint8_t*) &intValue, sizeof(intValue))` where intValue = 0.

Will integer 0 be encoded in a different form?

Reply:
It would be encoded in the expected way, but my concern is more whether it's possible for 0 to be intentionally stored at this key as opposed to representing a missing key. 

Further, there are multiple ways that `intValue` could be 0, and I wasn't sure which one we were checking for. Specifically, if `kv.value.size() == 0`, then intValue would remain 0, but it would also be 0 if the value was present and the encoding of 0.

Reply:
It is possible the value is 0 for two scenarios:
1) The ops key is never set in the workload. It will be the default value configured in the setup;
2) The value set for the ops key happen to be 0 (although unlikely). 

I will remove this if statement.

I can also make other changes to make `intValue == 0` to represent the missing of ops key, but it will involves changes to the actual workload (say all values set to ops key must be > 0). I don't prefer this solution because it will lose the ability to test the edge case (i.e., when intValue = 0), which is capable in the current test.

Reply:
It is unfortunately the latter case. All possible ops keys are set to 0 in the setup phase. 
An ops key can also be set to 0 in the workload-running phase.

So the end-result of 0 can be caused by either reasons:
1) the ops keys are never set, which is legal;
2) the ops keys are set to 0, which is legal;
3) the ops keys' set operation (i.e., AddValue) is missing in the database, which indicates a bug.